raiden-dapp/src/service-worker/worker/index.js

```diff
@@ -0,0 +1,96 @@
+/* istanbul ignore file */
+import { PrecacheController, PrecacheRoute } from 'workbox-precaching';
+import { registerRoute, setCatchHandler } from 'workbox-routing';
+import { ServiceWorkerMessages, ServiceWorkerAssistantMessages } from '../messages';
+import {
+  getPreservedPrecacheEntries,
+  preservePrecacheEntries,
+  deletePreservedPrecacheEntries,
+} from './database';
+import { isAnyClientAvailable, sendMessageToClients } from './clients';
+import { doesCacheExist, isCacheInvalid, deleteCache } from './cache';
+
+self.controller = new PrecacheController({ fallbackToNetwork: false });
+self.route = new PrecacheRoute(self.controller);
+
+async function update() {
+  await deleteCache();
+  await deletePreservedPrecacheEntries();
+  await sendMessageToClients.call(this, ServiceWorkerMessages.RELOAD_WINDOW);
+  await this.registration.unregister();
+}
+
+async function verifyCacheValidity() {
+  if (await isCacheInvalid.call(this)) {
+    sendMessageToClients.call(this, ServiceWorkerMessages.CACHE_IS_INVALID);
+  }
+}
+
+async function onInstall(event) {
+  const cacheExists = await doesCacheExist();
+  const preservedPrecacheEntries = await getPreservedPrecacheEntries();
+
+  if (!cacheExists) {
+    this.shouldUpdate = true;
+    this.precacheEntries = self.__WB_MANIFEST;
+    this.controller.addToCacheList(this.precacheEntries);
+    this.controller.install(event);
+  } else if (cacheExists && preservedPrecacheEntries) {
+    this.shouldUpdate = false;
+    this.precacheEntries = preservedPrecacheEntries;
+    this.controller.addToCacheList(this.precacheEntries);
+  } else {
+    this.installError = new Error('Cache given, but precache entries are missing!');
+  }
+}
+
+async function onActivate(event) {
+  if (this.shouldUpdate) {
+    await preservePrecacheEntries(this.precacheEntries);
+    this.controller.activate(event);
+  }
+
+  await this.clients.claim();
+
+  if (this.installError) {
+    sendMessageToClients.call(this, ServiceWorkerMessages.INSTALLATION_ERROR, this.installError);
+  }
+}
+
+async function onMessage(event) {
+  if (!event.data) return;
+
+  switch (event.data) {
+    case ServiceWorkerAssistantMessages.UPDATE:
+      await update.call(this);
+      break;
+
+    case ServiceWorkerAssistantMessages.VERIFY_CACHE:
+      await verifyCacheValidity.call(this);
+      break;
+
+    default:
+      break;
+  }
+}
+
+async function onRouteError() {
+  if (await isCacheInvalid.call(this)) {
+    if (await isAnyClientAvailable.call(this)) {
+      sendMessageToClients.call(this, ServiceWorkerMessages.CACHE_IS_INVALID);
+    } else {
+      update.call(this);
+      // Note that the user now has to reload the page himself. Since there is
+      // no client available, nobody can reload the page automatically.
```

Comment:
Can we display a message in this case?

Reply:
How?
This is basically what the comment states. There is no client. So no window, no document, no nothing.
But maybe I'm overseeing something. :shrug:

Reply:
This was just a verification. If we cannot do something we cannot.